staging/src/k8s.io/apiserver/pkg/endpoints/discovery/root/discovery.go

```diff
@@ -0,0 +1,88 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package root
+
+import (
+	"net/http"
+
+	restful "github.com/emicklei/go-restful/v3"
+
+	apidiscoveryv2beta1 "k8s.io/api/apidiscovery/v2beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apiserver/pkg/endpoints/discovery/aggregated"
+	"k8s.io/apiserver/pkg/endpoints/handlers/negotiation"
+	genericfeatures "k8s.io/apiserver/pkg/features"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+)
+
+const APIGroupPrefix = "/apis"
+
+type RootManager interface {
+	WebService() *restful.WebService
+}
+
+// RootAPIsHandler creates a webservice serving /apis
+type RootAPIsHandler struct {
+	serializer                      runtime.NegotiatedSerializer
+	discoveryGroupManager           http.Handler
+	aggregatedDiscoveryGroupManager http.Handler
+}
+
+func NewRootAPIsHandler(discoveryGroupManager, aggregatedDiscoveryGroupManager http.Handler) *RootAPIsHandler {
+	scheme := runtime.NewScheme()
+	apidiscoveryv2beta1.AddToScheme(scheme)
+	codecs := serializer.NewCodecFactory(scheme)
+
+	return &RootAPIsHandler{
+		serializer:                      codecs,
+		discoveryGroupManager:           discoveryGroupManager,
+		aggregatedDiscoveryGroupManager: aggregatedDiscoveryGroupManager,
+	}
+}
+
+func (s *RootAPIsHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+	if utilfeature.DefaultFeatureGate.Enabled(genericfeatures.AggregatedDiscoveryEndpoint) {
+		mediaType, _ := negotiation.NegotiateMediaTypeOptions(req.Header.Get("Accept"), s.serializer.SupportedMediaTypes(), aggregated.DiscoveryEndpointRestrictions)
+		// mediaType.Convert looks at the request accept headers and is used to control whether the discovery document will be aggregated.
+		if aggregated.IsAggregatedDiscoveryGVK(mediaType.Convert) {
+			s.aggregatedDiscoveryGroupManager.ServeHTTP(resp, req)
+			return
+		}
+	}
+	s.discoveryGroupManager.ServeHTTP(resp, req)
+}
+
+func (s *RootAPIsHandler) restfulHandle(req *restful.Request, resp *restful.Response) {
+	s.ServeHTTP(resp.ResponseWriter, req.Request)
+}
+
+// WebService returns a webservice serving api discovery.
+func (s *RootAPIsHandler) WebService() *restful.WebService {
+	mediaTypes, _ := negotiation.MediaTypesForSerializer(s.serializer)
+	ws := new(restful.WebService)
+	ws.Path(APIGroupPrefix)
+	ws.Doc("get available API versions")
+	ws.Route(ws.GET("/").To(s.restfulHandle).
+		Doc("get available API versions").
+		Operation("getAPIVersions").
+		Produces(mediaTypes...).
+		Consumes(mediaTypes...).
+		Writes(metav1.APIGroupList{}))
```

Comment:
Why does Writes reference the old API type and not the new one?

Reply:
It's the same endpoint as what the old types are using `/apis`. The switch for the new API is based on the Accept header and not the endpoint.